### ISSUE TITLE
Suggestion: default to not sending event if there is an error in the modules callback

### ISSUE THREAD

**Issue by benparsons** [matrix-org/synapse#10671]
A developer working on a new project using the modules system reports:

> if the module fails for some reason (e.g. with a runtime error in the Python code), the event is let through. This is worrying as it means that we don't have control on what the module can block. Is there a way to turn this around and have the module block the event by default in case of error?

The problem here is clearly that *outside* of the module interface it's not known which event types needed to be handled by the module. Is there a way we can flag events which have *not* been handled by the module callback and fail them?

**Comment by babolivier** [comment]
I'm not convinced the proposed solution is the right one. Modules are potentially untrusted third-party code and Synapse call them as such. So if the module is faulty, we want to prevent it from impacting the homeserver, since e.g. a faulty module implementing `check_event_for_spam` would then a) prevent any event from being sent via the C-S API and b) soft-fail any event received over federation, meaning a loss of data from the user pov which wouldn't be easily recoverable. So in my opinion the right way to fix this issue is more testing and reliability on the module's side, rather than making Synapse more vulnerable to failures from third-party code. But I'm happy to debate this if people have opposing opinions.

As a sidenote, `check_event_for_spam` doesn't currently have the error handling I'm referring to, so any failure from a module implementing it will fail the request, but we should probably fix this sooner rather than later.

**Comment by babolivier** [comment]
> The problem here is clearly that outside of the module interface it's not known which event types needed to be handled by the module.

It depends on the callback, if you're talking about `check_event_for_spam`, then the module is expected to handle all events (where by "handle" I also mean drop them for types it's not interested in).

**Comment by babolivier** [comment]
After discussing this internally it looks like the team's opinion is that we should fail the request if the module encounters an error. However, note that `check_event_for_spam` currently doesn't have this kind of error handling/protection, so I'm a bit unsure why this is showing up.

**Comment by erikjohnston** [comment]
Yeah, a quick look at the code suggests if `check_event_for_spam` fails by raising an exception the event should be rejected. Is that what they mean here?
